Potential regression due to healthcheck wait strategy change?

**Expected Behaviour**

We have a docker compose stack that we manage through testcontainers. Previous to v10.11.0, the stack came up normally.

**Actual Behaviour**

With v10.11.0, the stack fails to come up with a health check timeout.

**Testcontainer Logs**

```
Error: Health check not healthy after 60000ms

    at Timeout._onTimeout (/home/runner/work/monorepo/monorepo/node_modules/.pnpm/testcontainers@10.11.0/node_modules/testcontainers/src/wait-strategies/health-check-wait-strategy.ts:18:16)
```

**Steps to Reproduce**
1. Setup a docker compose file with health checks.
2. Call `await new DockerComposeEnvironment(composeFilePath, 'docker-compose.yml').withWaitStrategy('foo-1', Wait.forHealthCheck()).up();`
3. Observe the timeout.

**Environment Information**
- Operating System:
- Docker Version:
- Node version:
- Testcontainers version: v10.11.0

This is still a problem with the latest testcontainers release.

Doing some more debugging, it seems like maybe it's a race condition of sorts? I see that the containers become healthy well before testcontainers begins checking the event stream.

Here's the relevant logs with timestamps:

```
2024-09-24T06:23:14.500Z testcontainers:compose Container testcontainers-353d31fc9278-risk-1  Creating
2024-09-24T06:23:14.534Z testcontainers:compose Container testcontainers-353d31fc9278-risk-1  Created
2024-09-24T06:23:26.682Z testcontainers:compose Container testcontainers-353d31fc9278-risk-1  Starting
2024-09-24T06:23:26.995Z testcontainers:compose Container testcontainers-353d31fc9278-risk-1  Started
2024-09-24T06:23:26.996Z testcontainers:compose Container testcontainers-353d31fc9278-risk-1  Waiting
2024-09-24T06:23:26.996Z testcontainers:compose Container testcontainers-353d31fc9278-risk-1  Waiting
2024-09-24T06:23:28.496Z testcontainers:compose Container testcontainers-353d31fc9278-risk-1  Healthy
2024-09-24T06:23:28.497Z testcontainers:compose Container testcontainers-353d31fc9278-risk-1  Healthy
2024-09-24T06:23:38.555Z testcontainers [DEBUG] [dd5521f308bf] Getting container by ID...
2024-09-24T06:23:38.555Z testcontainers [DEBUG] [dd5521f308bf] Got container by ID
2024-09-24T06:23:38.555Z testcontainers [DEBUG] [dd5521f308bf] Inspecting container...
2024-09-24T06:23:38.568Z testcontainers [DEBUG] [dd5521f308bf] Inspected container
2024-09-24T06:23:38.568Z testcontainers [DEBUG] [dd5521f308bf] Fetching container logs...
2024-09-24T06:23:38.570Z testcontainers [DEBUG] [dd5521f308bf] Demuxing stream...
2024-09-24T06:23:38.570Z testcontainers [DEBUG] [dd5521f308bf] Demuxed stream
2024-09-24T06:23:38.570Z testcontainers [DEBUG] [dd5521f308bf] Fetched container logs
2024-09-24T06:23:38.570Z testcontainers [DEBUG] [dd5521f308bf] Waiting for container to be ready...
2024-09-24T06:23:38.570Z testcontainers [DEBUG] [dd5521f308bf] Waiting for health check...
2024-09-24T06:23:38.570Z testcontainers [DEBUG] [dd5521f308bf] Fetching event stream...
2024-09-24T06:23:38.668Z testcontainers [DEBUG] [dd5521f308bf] Fetched event stream...
24T06:24:38.669Z testcontainers [ERROR] [dd5521f308bf] Health check not healthy after 60000ms
24T06:24:38.669Z testcontainers [ERROR] [dd5521f308bf] Container failed to be ready: Error: Health check not healthy after 60000ms
2024-09-24T06:24:38.669Z testcontainers [DEBUG] [dd5521f308bf] Stopping container...
2024-09-24T06:24:39.041Z testcontainers [DEBUG] [dd5521f308bf] Stopped container
2024-09-24T06:24:39.041Z testcontainers [DEBUG] [dd5521f308bf] Removing container...
2024-09-24T06:24:39.054Z testcontainers [DEBUG] [dd5521f308bf] Removed container
```

--- a/packages/testcontainers/src/wait-strategies/health-check-wait-strategy.ts
+++ b/packages/testcontainers/src/wait-strategies/health-check-wait-strategy.ts
@@ -1,47 +1,31 @@
 import Dockerode from "dockerode";
 import { AbstractWaitStrategy } from "./wait-strategy";
-import { log } from "../common";
+import { IntervalRetry, log } from "../common";
 import { getContainerRuntimeClient } from "../container-runtime";
 
 export class HealthCheckWaitStrategy extends AbstractWaitStrategy {
   public async waitUntilReady(container: Dockerode.Container): Promise<void> {
     log.debug(`Waiting for health check...`, { containerId: container.id });
-
     const client = await getContainerRuntimeClient();
-    const containerEvents = await client.container.events(container, ["health_status"]);
 
-    return new Promise((resolve, reject) => {
-      const timeout = setTimeout(() => {
-        const message = `Health check not healthy after ${this.startupTimeout}ms`;
+    const status = await new IntervalRetry<string | undefined, Error>(100).retryUntil(
+      async () => (await client.container.inspect(container)).State.Health?.Status,
+      (healthCheckStatus) => healthCheckStatus === "healthy" || healthCheckStatus === "unhealthy",
+      () => {
+        const timeout = this.startupTimeout;
+        const message = `Health check not healthy after ${timeout}ms`;
         log.error(message, { containerId: container.id });
-        containerEvents.destroy();
-        reject(new Error(message));
-      }, this.startupTimeout);
+        throw new Error(message);
+      },
+      this.startupTimeout
+    );
 
-      const onTerminalState = () => {
-        clearTimeout(timeout);
-        containerEvents.destroy();
-        log.debug(`Health check wait strategy complete`, { containerId: container.id });
-      };
+    if (status !== "healthy") {
+      const message = `Health check failed: ${status}`;
+      log.error(message, { containerId: container.id });
+      throw new Error(message);
+    }
 
-      containerEvents.on("data", (data) => {
-        const parsedData = JSON.parse(data);
-
-        const status =
-          parsedData.status.split(":").length === 2
-            ? parsedData.status.split(":")[1].trim() // Docker
-            : parsedData.HealthStatus; // Podman
-
-        if (status === "healthy") {
-          resolve();
-          onTerminalState();
-        } else if (status === "unhealthy") {
-          const message = `Health check failed: ${status}`;
-          log.error(message, { containerId: container.id });
-          reject(new Error(message));
-          onTerminalState();
-        }
-      });
-    });
+    log.debug(`Health check wait strategy complete`, { containerId: container.id });
   }
 }

